### PR TITLE
Sprint 1 module village with location filter by name and level

### PR DESCRIPTION
**Pull request by ayocodingit**
# overview

- add redis for caching data villages
- search by name and level

cc: @jabardigitalservice/jds-backend 

## Evidence
- title: village with location filter by name and level
- project: Desa Digital
- participants: @ayocodingit @sandisunandar99 @tukangremot @rachadiannovansyah

### CODE REVIEW

**Review by rachadiannovansyah — APPROVED** [review]

File: src/modules/villages/village_handler.ts [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
 
 const router = express.Router()
 
 router.get(
   '/v1/villages/list-with-location',
-  async (req: Request, res: Response, next: NextFunction) => {
+  async (req: Request<never, never, never, Entity.RequestQuery>, res: Response, next: NextFunction) => {
     res.status(httpStatus.OK).json({
-      data: await Service.findAllWithLocation()
+      data: await Service.findAllWithLocation(req.query)
```
Comment (sandisunandar99) [inline src/modules/villages/village_handler.ts:12]: #ASK

for format output json is any discussion before? 
this is my reading article maybe you can considering that ,, 😄 
[https://medium.com/@bojanmajed/standard-json-api-response-format-c6c1aabcaa6d](https://medium.com/@bojanmajed/standard-json-api-response-format-c6c1aabcaa6d)

File: src/modules/villages/village_service.ts [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+  export const findAllWithLocation = async (requestQuery: Entity.RequestQuery): Promise<Entity.ResponseFindAllWithLocation[]> => {
+    let data: Entity.ResponseFindAllWithLocation[]
+    if (!await redis.get('find_all_with_location')) {
+      const items = await Repository.findAllWithLocation()
+      await redis.set('find_all_with_location', JSON.stringify(responseFindAllWithLocation(items)))
+    }
+
+    data = JSON.parse((await redis.get('find_all_with_location')))
+
+    if (requestQuery.name) data = data.filter(i => i.name.toLowerCase() === requestQuery.name.toLowerCase())
```
Comment (firmanJS) [inline src/modules/villages/village_service.ts:15]: filter is better in querying only, not object manipulation

Comment (firmanJS) [reply src/modules/villages/village_service.ts:15]: ref : https://github.com/jabardigitalservice/pikobar-pelaporan-backend/blob/master/services/map.js

Comment (ayocodingit) [reply src/modules/villages/village_service.ts:15]: better filter query database or data caching redis ?

Comment (firmanJS) [reply src/modules/villages/village_service.ts:15]: jadi gini, jika query nya ada filter baiknya key redis nya ada filter by query juga contoh key `find_all_with_location_nam_filter`

Comment (ayocodingit) [reply src/modules/villages/village_service.ts:15]: nanti jadi banyak key di redis nya yah

Comment (firmanJS) [reply src/modules/villages/village_service.ts:15]: wkwk siap gas

**Review by firmanJS — APPROVED** [review]

**Review by tukangremot — APPROVED** [review]

**Pull request merged**